### ISSUE TITLE
[BUG-230919] Alpha mode setting incorrectly for untextured objects

### ISSUE THREAD

**Issue by AtlasLinden** [secondlife/viewer#1366]
From jira-archive-internal created by [sl-service-account](https://github.com/sl-service-account): secondlife/jira-archive-internal#159839  

## What just happened?
Textures applied to blank untextured objects (uploaded mesh) are setting PRIM_ALPHA_MODE to BLEND, regardless of alpha channel of the texture.  On rezzing textures, instead of appearing solid grey, the face will sometimes appear invisible, and at times will suffer from alpha flicker before the object is selected in Build to "remind" the viewer it has no alpha channel.  Alpha mode has been verified by script, not just visually: it returns 1 (BLEND) instead of 0 (NONE).  Objects set to an alpha texture and then to a solid texture also appear to now be retaining the BLEND setting, instead of switching mode correctly to reflect the texture type.

## What were you doing when it happened?
I applied solid no alpha channel textures to mesh (untextured on upload) objects.  After mentioning the issue, I discovered I was not the only person who has seen this behavior. 

## What were you expecting to happen instead?
Previously, texture alpha mode would be (mostly) set by the texture- if it had no alpha, it would be set to PRIM_ALPHA_MODE_NONE (0), and would be set to BLEND (1) only if the texture had an alpha channel.  While it had some failure (NONE would be forced on alpha textures if materials were applied first- it appears to set alpha mode correctly now when materials are added first- BLEND or NONE as required by the texture), it was much easier to change the alpha to the correct mode on a texture with an alpha channel- the dropdown does not exist with textures without alpha.  Setting the object to a texture with alpha and then without still retains the BLEND setting. 

# Other information
This is a relatively new bug- early June 2021 is the first time I noticed it without figuring out what was really going on, I believe I've found it set incorrectly in an object initially textured in early April.  An item that I think was textured in the same manner (possibility exists I added materials first though- adding materials first sets mode to NONE with solid textures) in February does not seem to have the same issue.  Objects textured initially between February and April were rigged mesh that did texture with the correct mode: rigged mesh NOW seems to have the same issue as unrigged mesh with alpha modes, but I do not know if they did then. 

## Attachments
- [alpha glitch.png](https://jira-attachments.secondlife.io/BUG/122375.png) 2021-12-11 23:50:53

### Related
- [ ] https://github.com/secondlife/jira-archive/issues/16303
- [ ] https://github.com/secondlife/jira-archive-internal/issues/156809
- [ ] https://github.com/secondlife/jira-archive/issues/7936

<details>
<summary>Original Jira Fields</summary>

| Field          | Value         |
| ------------- | ------------- |
| Issue         | BUG-230919 |
| Summary       | Alpha mode setting incorrectly for untextured objects |
| Type          | Bug |
| Priority      | Unset |
| Status        | Needs More Info |
| Resolution    | Unresolved |
| Reporter      | Allegory Malaprop (allegory.malaprop) |
| Created at    | 2021-07-03T07:01:37Z |
| Updated at    | 2022-11-20T21:54:46Z |

```
{
  'Build Id': 'unset',
  'Business Unit': ['Platform'],
  'Date of First Response': '2021-08-31T13:35:44.401-0500',
  "Is there anything you'd like to add?": "This is a relatively new bug- early June 2021 is the first time I noticed it without figuring out what was really going on, I believe I've found it set incorrectly in an object initially textured in early April.  An item that I think was textured in the same manner (possibility exists I added materials first though- adding materials first sets mode to NONE with solid textures) in February does not seem to have the same issue.  Objects textured initially between February and April were rigged mesh that did texture with the correct mode: rigged mesh NOW seems to have the same issue as unrigged mesh with alpha modes, but I do not know if they did then.",
  'ReOpened Count': 0.0,
  'Severity': 'Unset',
  'System': 'SL Viewer',
  'Target Viewer Version': 'viewer-development',
  'What just happened?': 'Textures applied to blank untextured objects (uploaded mesh) are setting PRIM_ALPHA_MODE to BLEND, regardless of alpha channel of the texture.  On rezzing textures, instead of appearing solid grey, the face will sometimes appear invisible, and at times will suffer from alpha flicker before the object is selected in Build to "remind" the viewer it has no alpha channel.  Alpha mode has been verified by script, not just visually: it returns 1 (BLEND) instead of 0 (NONE).  Objects set to an alpha texture and then to a solid texture also appear to now be retaining the BLEND setting, instead of switching mode correctly to reflect the texture type.',
  'What were you doing when it happened?': 'I applied solid no alpha channel textures to mesh (untextured on upload) objects.  After mentioning the issue, I discovered I was not the only person who has seen this behavior.',
  'What were you expecting to happen instead?': 'Previously, texture alpha mode would be (mostly) set by the texture- if it had no alpha, it would be set to PRIM_ALPHA_MODE_NONE (0), and would be set to BLEND (1) only if the texture had an alpha channel.  While it had some failure (NONE would be forced on alpha textures if materials were applied first- it appears to set alpha mode correctly now when materials are added first- BLEND or NONE as required by the texture), it was much easier to change the alpha to the correct mode on a texture with an alpha channel- the dropdown does not exist with textures without alpha.  Setting the object to a texture with alpha and then without still retains the BLEND setting.',
  'Where': 'http://maps.secondlife.com/secondlife/Sandbox+Pristina/46/170/27\r\n\r\nAlso witnessed on other sims.',
}
```

</details>

**Comment by AtlasLinden** [comment]
_Dan Linden commented at 2021-08-31T18:35:44Z, updated at 2021-08-31T21:14:21Z_

When I upload a mesh object,
add this script to it,

```Java

default
{
    touch_start(integer total_number)
    {
        list my_list = llGetPrimitiveParams([ PRIM_ALPHA_MODE, 0 ]); 
        llOwnerSay("alpha_mode=" + llList2String(my_list,0) + ".  0=NONE, 1=BLEND, 2=MASK, 3=EMISSIVE");
        llOwnerSay("mask_cutoff=" + llList2String(my_list,1));
    }
}
```

and click it,
I see "alpha_mode=1"
You assert that it should be "alpha_mode=0"

Still investigating.

**Comment by AtlasLinden** [comment]
_Dan Linden commented at 2021-09-24T14:30:45Z_

I put an alpha texture on a box, then set the alpha mode to None.
I changed the texture to default (plywood) and checked the alpha mode with the script. The script reported the alpha mode was Blend.

**Comment by AtlasLinden** [comment]
_Dan Linden commented at 2021-09-28T21:11:24Z, updated at 2021-09-28T21:13:06Z_ **[PRIVATE]**

Tested with an old viewer circa March 2020, and saw the same result on a new box, "alpha_mode=1, BLEND". Since the original reporter said this bug started around June 2021, this bug does not seem to be caused by the viewer.

**Comment by AtlasLinden** [comment]
_Dan Linden commented at 2021-09-29T16:56:40Z, updated at 2022-02-01T17:09:18Z_

Hi  Allegory,

I've tested this with an old viewer from 2016 and see the same results, a new default box has a PRIM_ALPHA_MODE=1.

We don't believe anything has changed on the server side that would affect the alpha mode of a new object.

Do you have some other proof that a new object was internally flagged as PRIM_ALPHA_MODE 0 in the past?

Can you provide a location where we can see the alpha flicker before the object is selected?

There are bugs about alpha mode that changed sometime after object creation; (edit: BUG-230211, not BUG-23021) BUG-11333 BUG-8715. Could what you are seeing be explained by one of those?

Thank you

**Comment by AtlasLinden** [comment]
_Allegory Malaprop commented at 2021-10-01T02:53:40Z_

I haven't seen the alpha flicker or invisible rezzing with objects believing they were set incorrectly recently- but it was not an entirely consistent error when I was seeing it.  It was only a brief window in which it was visible, so I'm now assuming it was a viewer side rendering issue that has since been fixed- it sounds potentially related to the 2 visible issues (BUG-23021 is not in an area of the Jira I am allowed to access).

However something DID briefly change in that window with alpha modes that has since changed back to old behavior.  Adding materials (normal and specular textures) to an object, mesh or prim, that has a solid no alpha texture, sets its alpha mode to NONE explicitly- adding a texture with an alpha then to the object displays it without the alpha.  (I think it's always been that way- aside from when I filed this report.)  When I saw the visible texture alpha rezzing faults, adding materials first and then adding a transparent texture displayed the alpha mode as BLEND correctly, as mentioned in the "What were you expecting to happen instead?" section.

**Comment by AtlasLinden** [comment]
_Allegory Malaprop commented at 2021-12-12T04:52:05Z_

Something related to this bug appears to be back.  I've only seen it happen on my Linden Home sim, Leech Lake- where the previously mentioned bug was also happening very obviously the first time.

This time, however, some solid textured faces (at random!  only some!  in the same object, identical settings!) are entirely disappearing even when textures are fully rezzed.  Object uploaded day before yesterday.

![alpha glitch.png](alpha glitch.png)

**Comment by AtlasLinden** [comment]
_Whirly Fizzle commented at 2021-12-18T16:00:08Z_

Hiya Allegory,

I suspect in your last comment you are seeing BUG-9439.

Tl/dr: in the location where you see this bug, there are too many mesh triangles present within a small area which causes the viewer to run out of memory for that spatial partition node.
This is technically not a bug, it is deliberate for performance reasons.
The debug setting controlling this is"RenderMaxNodeSize" and defaults to 64MB per node.
When you notice the baubles not rendering again, if you **temporarily** raise the debug RenderMaxNodeSize and the baubles then render, it is definitely BUG-9439.
Do not leave RenderMaxNodeSize at higher then the default setting - the viewer will suffer from poor performance & may even crash in certain scenes.

**Comment by AtlasLinden** [comment]
_Dan Linden commented at 2022-02-01T17:35:50Z_ **[PRIVATE]**

LINDEN ONLY COMMENT

Deer head is at http://maps.secondlife.com/secondlife/Leech+Lake/241/138/49

**Comment by AtlasLinden** [comment]
_Dan Linden commented at 2022-10-14T17:43:19Z_

Are you still seeing this? Does Whirly's suggested workaround fix this?

**Comment by AtlasLinden** [comment]
_Pazako Karu commented at 2022-11-20T21:54:47Z_

There is a known issue where a non-alpha png uploads as having an alpha layer, even if it is blank. As with other blending things, it may then get clipped behind other alpha textures depending on view angle. If your non alpha texture can be set to blending, this is what happened.

The best solution against this is to save your files as tga and upload as tga. You can also use jpg, but that will remove alpha on everything. Tga just works properly.

**Comment by AtlasLinden** [comment]
We have a report that this issue has recurred multiple times on [Merry Land](https://support-tools.agni.lindenlab.com/gridtool/region/c4db6dc5-d156-43d7-963b-7f9453c60e0f/) in previous weeks.

**Comment by AtlasLinden** [comment]
One more report on [[Chaisuki](https://support-tools.agni.lindenlab.com/gridtool/region/f0d2be50-1bae-45a3-ae16-0ec06c4474b1/)].

> Since Monday of this week (2024-03-18) many many of our plants that we have rezzed have been unrecognisable with large black areas in place of their textures.  When we check the object in edit the "Alpha Masking" has changed from "Blending" as it should be to "None".

This comment on a separate archived issue is the best lead I can find:
https://github.com/secondlife/jira-archive-internal/issues/156809#issuecomment-1931192694
